doc/source/raysgd/raysgd_pytorch.rst

```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
         num_replicas=100,
         use_gpu=True)
 
-See the documentation on the PyTorchTrainer here: :ref:`ref-pytorch-trainer`. We'll look at the training APIs next.
+See the documentation on the PyTorchTrainer here: :ref:`ref-pytorch-trainer`.
 
-Training APIs
--------------
 
-Now that the trainer is constructed, you'll naturally want to train the model.
+.. _raysgd-custom-training:
 
-.. code-block:: python
+Custom Training and Validation (Operators)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    trainer.train()
+``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexibility similar to using PyTorch natively.
+This is done via the :ref:`ref-pytorch-operator` interface.
+
+For both training and validation, there are two granularities that you can provide customization - per epoch or per batch. These correspond to ``train_batch``,
+``train_epoch``, ``validate``, and ``validate_batch``. Other useful methods to override include ``setup``, ``save`` and ``restore``. You can use this class
+to manage state (like a classifier neural network for calculating inception score, or a heavy tokenizer).
 
-This takes one pass over the training data.
+Providing a custom operator is necessary if creator functions return multiple models, optimizers, or schedulers.
 
-To run the model on the validation data passed in by the ``data_creator``, you can simply call:
+Below is a partial example of overriding ``train_batch``.
 
 .. code-block:: python
 
-    trainer.validate()
+    from ray.util.sgd.pytorch import TrainingOperator
+
+    class GANOperator(TrainingOperator):
+        def setup(self, config):
+            """Custom setup for this operator.
+
+            Args:
+                config (dict): Custom configuration value to be passed to
+                    all creator and operator constructors. Same as ``self.config``.
+            """
+            self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+        def train_batch(self, batch, batch_info):
+            """Trains on one batch of data from the data creator.
+
+            Args:
+                batch: One item of the validation iterator.
+                batch_info (dict): Information dict passed in from ``train_epoch``.
+
+            Returns:
+                A dict of metrics. Defaults to "loss" and "num_samples",
+                    corresponding to the total number of datapoints in the batch.
+            """
+            real_label = 1
+            fake_label = 0
+            discriminator, generator = self.models
+            optimD, optimG = self.optimizers
+
+            discriminator.zero_grad()
```

Comment:
comments with high-level descriptions of what's happening in each of these blocks would be helpful

Reply:
fixed.